### PR TITLE
p2p: add Channels to NodeInfo and don't send for unknown channels

### PR DESCRIPTION
**Pull request by ebuchman**
Solves https://github.com/tendermint/tendermint/issues/957

### CODE REVIEW

File: p2p/types.go [hunk 1]
```diff
@@ -51,6 +53,24 @@ func (info *NodeInfo) CompatibleWith(other *NodeInfo) error {
...
+	if len(info.Channels) == 0 {
+		return nil
+	}
+
+	// for each of our channels, check if they have it
+	found := false
+	for _, ch1 := range info.Channels {
+		for _, ch2 := range other.Channels {
+			if ch1 == ch2 {
+				found = true
```
Comment (melekes) [inline p2p/types.go]: should we `break` after?

File: node/node.go [hunk 2]
```diff
@@ -551,10 +551,17 @@ func (n *Node) makeNodeInfo(pubKey crypto.PubKey) *p2p.NodeInfo {
 		Moniker: n.config.Moniker,
 		Network: n.genesisDoc.ChainID,
 		Version: version.Version,
+		Channels: []byte{
+			bc.BlockchainChannel,
+			cs.StateChannel, cs.DataChannel, cs.VoteChannel, cs.VoteSetBitsChannel,
+			mempl.MempoolChannel,
+			evidence.EvidenceChannel,
+			p2p.PexChannel,
```
Comment (melekes) [inline node/node.go]: what if `pex` is disabled?

**Comment by melekes** [comment]
> Then if a peer sends us a msg for a channel we don't know about, we can mark them as bad and disconnect :)

what about this part?

**Comment by ebuchman** [comment]
The old code wasn't actually doing what we wanted. Force pushed  the fix. PTAL

**Comment by ebuchman** [comment]
> what about this part?

This is implemented already since if we receive a msg for an unknown channel we stopPeerForError.

**Comment by ebuchman** [comment]
Saving that for https://github.com/tendermint/tendermint/issues/1148

File: p2p/node_info.go [hunk 3]
```diff
@@ -71,6 +90,24 @@ func (info NodeInfo) CompatibleWith(other NodeInfo) error {
...
+		return nil
+	}
+
+	// for each of our channels, check if they have it
+	found := false
+	for _, ch1 := range info.Channels {
+		for _, ch2 := range other.Channels {
+			if ch1 == ch2 {
+				found = true
+				break // only need one
```
Comment (melekes) [inline p2p/node_info.go]: I think the intention here was to break from outer loop, no?

**Pull request merged**